### ISSUE TITLE
Allow accessing the asset value loader from `OpExecutionContext`

### ISSUE THREAD

**Issue by danielgafni** [dagster-io/dagster#15452]
### What's the use case?

Currently `Definitions.load_asset_value` can be used to load assets. 
But because `Definitions` is the root object in the Dagster project hierarchy, accessing it inside an op would cause a circular dependency. 
Some typecheckers like `pyright` are very strict about circular dependencies, even if they happen at runtime. 
Currently I need to use something like this

```python
definitions = import_module("my_module").definitions
assert isinstance(definitions, Definitions)
```

to access `definitions` in an `op`. 

It would be nice if the `definitions` or the `asset_value_loader` would be provided by Dagster and accessible via `context`. 

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sryza** [comment]
I agree this would be a good change. If you're interested in making it I'd be happy to review a PR.

A warning: I tried this one time and it was gnarlier than I expected, because it requires finding a way to thread Definitions down a bunch of different code paths. Also, we need to figure out what to do in the case that someone executes their op using `materialize` - i.e. there is no user-created `Definitions` object.

**Comment by jrstats** [comment]
Hi all, I think I have another use case for this (unless I am missing something obvious).

I have several assets that each get country level data from different sources in a different way, but transform them into the same format (e.g. `country_usa`, `country_gbr`, etc.).

I then wanted to create an `all_countries` asset (partitioned by country) that is essentially a concatenation of the country assets, and updates the country partition when the country asset is updated (via a `multi_asset_sensor`, for example).

So the `all_countries` asset would be defined as:

``` py

@asset(...)
def all_countries(context: AssetExecutionContext) -> pd.DataFrame:
    country_code = context.partition_key
    df = context.load_asset_value(f"country_{country_code}")
    return df
``` 

I have tried the `import_module` approach described by @danielgafni, but am still getting circular import errors.

``` py
import pandas as pd
from dagster import asset, AssetExecutionContext
from importlib import import_module

## neither can get the defs object from mymodule/__init__.py
defs = import_module("my_module.__init__").defs # DagsterImportError (most likely due to a circular import)
defs = import_module("my_module").defs # AttributeError: partially initialized module 'graphs' has no attribute 'defs' (most likely due to a circular import)

@asset
def all_countries(context: AssetExecutionContext) -> pd.DataFrame:
    country_code = context.partition_key
    df = defs.load_asset_value(f"country_{country_code}")
    return df
```

I was then hoping to create an `AssetValueLoader` instance but cant get the upstream defs from the context.
Is this the challenge that you refer to @sryza, to add something like `upstream_assets_defs` to the `AssetExecutionContext`? I am not too confident on how this is constructed?


The only way I have managed is to create a new def. I have to include all the upstream assets, as well as the resource dict, but this is defined somewhere else, so can be imported okay.

**Comment by danielgafni** [comment]
1. Try to move the import instead to the asset body 
2. Check if the circular import errors actually point to definitions.py or asset.py. If not, you can probably resolve them by restructuring your modules (e.g. separate the problematic code into a separate module).

**Comment by jrstats** [comment]
> Try to move the import inside the asset body

That is the one. Thanks very much @danielgafni :pray: 

``` py
import pandas as pd
from dagster import asset, AssetExecutionContext
from importlib import import_module

@asset
def all_countries(context: AssetExecutionContext) -> pd.DataFrame:
    country_code = context.partition_key
    defs = import_module("my_module").defs
    df = defs.load_asset_value(f"country_{country_code}")
    return df
```

**Comment by danielgafni** [comment]
By the way, it seems like currently it's impossible to express the asset dependencies you need with Dagster. 

You would need to map each downstream partition into exactly one upstream asset. That's not really possible. 

My advice would be to keep a single partitioned asset and simply execute different code based on partition_key value.

**Comment by jrstats** [comment]
I did consider that approach, but I came across the `multi_asset_sensor` example in the docs and I'm pretty sure meets my needs.

I've got a set of `monitored_assets`, that I can pass into `multi_asset_sensor` and the `non_argument_deps` (for the UI) for the `all_countries` asset.

``` py

monitored_assets = AssetSelection.keys(asset_key1) | ...
job_name = "all_countries_job"

@asset(
    partitions_def=country_partition_def,
    non_argument_deps=set(key for key in monitored_assets._keys),
)
def all_countries(context: AssetExecutionContext) -> pd.DataFrame:
    ...

all_countries_job = define_asset_job(
    name=job_name, 
    selection=[all_countries], 
    partitions_def=country_partition_def
)

@multi_asset_sensor(job_name=job_name, monitored_assets=monitored_assets)
def country_sensor(context: MultiAssetSensorEvaluationContext):
    ...
```

**Comment by danielgafni** [comment]
Sure, but in this case the upstream assets will appear disconnected from the downstream asset, even tho it's actually dependent on them. You can't use `deps=upstream_assets` since this will make every partition dependent on every asset. 

This is fine, just not exactly ideal IMO. The sensor is alright and this setup will work, but I won't consider it idiomatic since it breaks your asset graph. I would prefer using `AutoMaterializationPolicy` over sensors, but that's also my personal preference.

**Comment by jrstats** [comment]
Yeah, I wanted to avoid every partition dependent on every upstream asset.

Having the list of assets as `non_argument_deps` or `deps` shows the connection within the asset graph but doesn't log a `LOADED_INPUT` event on any of the `deps` when materialising the `all_countries` asset. I am not super confident, but I thought that meant this was not the case:

> You can't use `deps=upstream_assets` since this will make every partition dependent on every asset.

I also prefer the `AutoMaterializationPolicy` over sensors, but can't work out how to trigger a run for the country partition based on the country asset (as we would quite like to have the individual country assets). And a sensor did give this capability.

**Comment by jrstats** [comment]
The intermediate assets are also useful in of themselves, unfortunately.

**Comment by danielgafni** [comment]
in this case you can reverse it and make them downstream to the combined asset?
